src/dump.c

```diff
@@ -453,6 +454,57 @@ static void jl_serialize_value_(jl_serializer_state *s, jl_value_t *v, int as_li
         ptrhash_put(&backref_table, v, (char*)HT_NOTFOUND + pos + 1);
     }
 
+    return 0;
+}
+
+static void jl_serialize_code_instance(jl_serializer_state *s, jl_code_instance_t *codeinst, int skip_partial_opaque) JL_GC_DISABLED
+{
+    if (jl_serialize_generic(s, (jl_value_t*)codeinst)) {
+        return;
+    }
+
+    int validate = 0;
+    if (codeinst->max_world == ~(size_t)0)
+        validate = 1; // can check on deserialize if this cache entry is still valid
+    int flags = validate << 0;
+    if (codeinst->invoke == jl_fptr_const_return)
+        flags |= 1 << 2;
+    if (codeinst->precompile)
+        flags |= 1 << 3;
+
+    int write_ret_type = validate || codeinst->min_world == 0;
+    if (write_ret_type && codeinst->rettype_const &&
+            jl_typeis(codeinst->rettype_const, jl_partial_opaque_type)) {
```

Comment:
If this happens, I think we should also see this same rettype_const object in `Method.roots` and possibly elsewhere, from when we compressed `slottypes`? We might want to improve that eventually, but seems like we need to just make this valid to store? What's inhibiting that currently?

Reply:
The PartialOpaque will have a reference to the `Method` object that we don't have a good identity for.